### PR TITLE
Fix/multiple requests listing

### PR DESCRIPTION
**Pull request by milohr**
The issue is that a request was being sent one after another for changes in the `page`, `pageSize`, `category`, `nameFilters`, `tags`, etc, fields.

So if a `page` and `tag` field changed closed in time, then two requests would be sent, resulting in repeated and unrelated mixed results, from both requests' replies.

My proposed solution is to use a single-shot timer to control multiple requests close in time. It has a timeout of `900 ms`, all requests within that time are ignored until the timer is timeout, then the request is sent with whatever arguments all those fields have at that moment.

### CODE REVIEW

**Comment by azubieta** [comment]
While timers can help here there is still a chance for the same issue to happen if the network connection is too slow.

@milohr I would go for encapsulating the request and response handler into an object (class) that could be disposed if a new one is made before the later is completed.

**Comment by probal31** [comment]
@milohr Searching apps by text multiple times keeps on appending the results.

For example, Search "Instagram" shows 2 results.
Again search with text "Instagra", now again 2 apps are appended to the list (total 4).
Again searching with text "Instagram" now has total 6 apps (2 unique apps, rest are duplicates).

**Comment by milohr** [comment]
> While timers can help here there is still a chance for the same issue to happen if the network connection is too slow.
> 
> @milohr I would go for encapsulating the request and response handler into an object (class) that could be disposed if a new one is made before the later is completed.

I thought about this queueing solution but it might present the following issue:

If I make a request such as:

`tags=[appimages]`
 then
 
```
 tags=[mauikit]
 nameFilter="index"
```

i will see results from the first request and then it would jump to the actual results of the second request. This is very likely to happen often due to how the StoreModel is designed: every change in any of the fields triggers a request, so to clear a "search" all the fields would end up triggering a request, meaning a big "queue"

Hope to have explained myself here.

An alternative solution, but requiering refactoring, is forcing the StoreModle API to make only request in a single package, not by every field change.

so instead of 
```
fieldChanged() -> request()
fieldChanged() -> request()
fieldChanged() -> request()
```

do
```
setField()
setField()
setField()
request()
```

**Review by azubieta — CHANGES_REQUESTED** [review]

File: src/models/categorymanager.h [hunk 1]
```diff
@@ -0,0 +1,22 @@
+#pragma once
+
+#include "../stores/apprepostore.h"
+#include "../stores/appimagehubstore.h"
+
+class CategoryManager : public QObject {
```
[the remark this answers is not in the document]
Comment (probal31) [reply src/models/categorymanager.h]: Updated the store manager implementation. Previously category manager and application manager was separate.
Combined those implementations to the store manager

File: src/models/applicationmanager.cpp [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+                                         QString pageSize, QList<QString> tags,
+                                         Store::Arch arch, Category *category) {
+
+    if ( category->categoryStore == Category::CategoryStore::APPIMAGEHUB ) {
+        // Invoke appimagehub api
+        m_appimagehubStore->getApplicationsByArch(categoriesFilter, nameFilter, sortMode, page, pageSize, tags, arch);
+
+        // On clicking a top level category (appimagehub category), fetch the apps from apprepo sub-categories as well
+        // which is there under the clicked top level category. So that all apps from appimagehub and apprepo are shown
+        // foreach (Category *item, category->categories) {
```
[the remark this answers is not in the document]
Comment (probal31) [reply src/models/applicationmanager.cpp]: Removed unused code

File: src/models/applicationmanager.cpp [hunk 3]
```diff
@@ -0,0 +1,50 @@
+#include "applicationmanager.h"
+
+ApplicationManager::ApplicationManager(QObject *parent)
```
[the remark this answers is not in the document]
Comment (probal31) [reply src/models/applicationmanager.cpp]: Updated the implementation and added a combined logic as store manager

File: src/models/applicationmanager.cpp [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    : QObject(parent)
+    , m_appimagehubStore(new AppImageHubStore(this))
+    , m_apprepoStore(new AppRepoStore("https://apprepo.de/rest/api", this))
```
[the remark this answers is not in the document]
Comment (probal31) [reply src/models/applicationmanager.cpp]: Added code to allow lazy loading

**Review by azubieta — APPROVED** [review]

File: tests/services/storemanager/teststoremanagerservice.cpp [hunk 5]
```diff
@@ -46,6 +46,48 @@ void TestStoreManagerService::testGetApplications()
     QVERIFY(applicationResult->applications.size() <= 10);
 }
 
+void TestStoreManagerService::testGetApprepoApplications()
```
Comment (azubieta) [inline tests/services/storemanager/teststoremanagerservice.cpp:49]: :+1: automatic testing is the way to great software :)

**Review by UriHerrera — APPROVED** [review]

**Pull request merged**